src/pages/blogs/how_shorebird_works.md

```diff
@@ -0,0 +1,229 @@
+for a function to use to execute. This allows us to effectively replace parts of
+your application at runtime without needing to compile new code on the device.
+
+## An interpreter for Dart
+
+Adding an interpreter to Dart was a challenge (working on compilers is hard, if
+you like compilers, we’re [hiring](https://shorebird.dev/jobs/)), and took us
+most of the last year. We didn’t try to build a particularly fancy interpreter
+for Dart (that had been attempted at Google multiple times before), but rather
+built something very simple. One of the challenges this creates is that
+Interpreters (and particularly our simple one) are very slow. In our case, our
+current (unoptimized) interpreter is about 100x slower than executing Dart AOT
+code on a CPU. Thankfully we had an insight early on that made us not have to
+care about interpreter speed.
+
+The insight that makes Shorebird possible is building a system that runs only
+changed Dart logic on the interpreter. Since the vast majority of the
+performance-critical Dart code in a Flutter program is typically the Flutter
+framework itself, and very few users ever change the Flutter framework, as long
+as we’re able to integrate our interpreter well and execute unchanged code on
+the CPU, essentially all of your application would end up running (at full
+speed) on the CPU and your program as a whole would show unchanged performance.
+This was our bet. Other than being extremely hard to make work, it paid off.
+
+Determining what parts of your program we could run on the CPU vs. interpreter
+was hard. To do this we invented a new phase of Dart compilation we called the
+"linker". The linker’s job is to analyze two (similar) Dart programs and find
+the maximal similarity between them and then decide what would be necessary to
+update in the first one in order to make it run like the second. We still have a
+[couple missing
+optimizations](https://github.com/shorebirdtech/shorebird/issues/1892) in this
+part of our system, but when it works well, developers see 99% of their code run
+on the CPU (even for large changes). Teaching the linker how to figure this out
+however required significant changes to Dart’s compiler toolchain.
+
+## Changes to Dart’s Toolchain
+
+We made many changes to Dart, including:
+
+1. Runtime changes to add our interpreter.
+2. Compiler changes teach the Dart compiler how to compile a program maximally
+   similar to a previous version of the same program.
+3. A new linker which can compare parts two versions of an app and decide if and
+   how they’ve changed.
+
+And much more -- too many to go through here, but I’ll give one example of a
+problem we recently solved.
+
+## Optimizing Dart’s constant pool
+
+In programming languages it’s typically to have "constants" which are variables
+that don’t change at runtime. These are often pre-computed during compile time,
+saved in a common space and shared throughout the program. E.g. if you have the
```

Comment:
```suggestion
saved in a common space and shared throughout the program. E.g., if you have the
```

Reply:
Does e.g. typically have a comma after it?

Reply:
According to the AP style guide

Reply:
🤯